File: MetaWearApiTest/DevicesTableViewController.m
/**
 * DevicesTableViewController.m
 * MetaWearApiTest
 *
 * Created by Stephen Schiffli on 7/29/14.
 * Copyright 2014-2015 MbientLab Inc. All rights reserved.
 *
 * IMPORTANT: Your use of this Software is limited to those specific rights
 * granted under the terms of a software license agreement between the user who
 * downloaded the software, his/her employer (which must be your employer) and
 * MbientLab Inc, (the "License").  You may not use this Software unless you
 * agree to abide by the terms of the License which can be found at
 * www.mbientlab.com/terms.  The License limits your use, and you acknowledge,
 * that the Software may be modified, copied, and distributed when used in
 * conjunction with an MbientLab Inc, product.  Other than for the foregoing
 * purpose, you may not use, reproduce, copy, prepare derivative works of,
 * modify, distribute, perform, display or sell this Software and/or its
 * documentation for any purpose.
 *
 * YOU FURTHER ACKNOWLEDGE AND AGREE THAT THE SOFTWARE AND DOCUMENTATION ARE
 * PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTY OF MERCHANTABILITY, TITLE,
 * NON-INFRINGEMENT AND FITNESS FOR A PARTICULAR PURPOSE. IN NO EVENT SHALL
 * MBIENTLAB OR ITS LICENSORS BE LIABLE OR OBLIGATED UNDER CONTRACT, NEGLIGENCE,
 * STRICT LIABILITY, CONTRIBUTION, BREACH OF WARRANTY, OR OTHER LEGAL EQUITABLE
 * THEORY ANY DIRECT OR INDIRECT DAMAGES OR EXPENSES INCLUDING BUT NOT LIMITED
 * TO ANY INCIDENTAL, SPECIAL, INDIRECT, PUNITIVE OR CONSEQUENTIAL DAMAGES, LOST
 * PROFITS OR LOST DATA, COST OF PROCUREMENT OF SUBSTITUTE GOODS, TECHNOLOGY,
 * SERVICES, OR ANY CLAIMS BY THIRD PARTIES (INCLUDING BUT NOT LIMITED TO ANY
 * DEFENSE THEREOF), OR OTHER SIMILAR COSTS.
 *
 * Should you have any questions regarding your right to use this Software,
 * contact MbientLab via email: hello@mbientlab.com
 */

#import "DevicesTableViewController.h"
#import "DeviceDetailViewController.h"
#import "MBProgressHUD.h"
#import <MetaWear/MetaWear.h>
#import <iOSDFULibrary/iOSDFULibrary-Swift.h>

@interface DevicesTableViewController () <DFUPeripheralSelector, LoggerDelegate, DFUServiceDelegate, DFUProgressDelegate>
@property (nonatomic) NSArray *devices;
@property (nonatomic) UIActivityIndicatorView *activity;
@property (nonatomic) MBProgressHUD *hud;
@property (nonatomic) MBLMetaWear *selected;

@property (weak, nonatomic) IBOutlet UISwitch *scanningSwitch;
@property (weak, nonatomic) IBOutlet UISwitch *metaBootSwitch;
@end

@implementation DevicesTableViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    self.activity = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleGray];
    self.activity.center = CGPointMake(95, 138);
    [self.tableView addSubview:self.activity];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [self setScanning:self.scanningSwitch.on];
}

- (void)viewDidDisappear:(BOOL)animated
{
    [super viewDidDisappear:animated];
    
    [self setScanning:NO];
}

- (void)setScanning:(BOOL)on
{
    if (on) {
        [self.activity startAnimating];
        if (self.metaBootSwitch.on) {
            [[MBLMetaWearManager sharedManager] startScanForMetaBootsAllowDuplicates:YES handler:^(NSArray *array) {
                self.devices = array;
                [self.tableView reloadData];
            }];
        } else {
            [[MBLMetaWearManager sharedManager] startScanForMetaWearsAllowDuplicates:YES handler:^(NSArray *array) {
                self.devices = array;
                [self.tableView reloadData];
            }];
        }
    } else {
        [self.activity stopAnimating];
        [[MBLMetaWearManager sharedManager] stopScan];
    }
}

- (IBAction)scanningSwitchPressed:(UISwitch *)sender
{
    [self setScanning:sender.on];
}

- (IBAction)metaBootSwitchPressed:(id)sender
{
    [[MBLMetaWearManager sharedManager] stopScan];
    // Wait a split second for any final callbacks to fire before starting up scanning again
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        self.devices = nil;
        [self.tableView reloadData];
        [self setScanning:self.scanningSwitch.on];
    });
}

- (IBAction)clearListPressed:(id)sender
{
    
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.devices.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    MBLMetaWear *cur = self.devices[indexPath.row];
    
    UILabel *uuid = (UILabel *)[cell viewWithTag:1];
    uuid.text = cur.identifier.UUIDString;
    
    UILabel *rssi = (UILabel *)[cell viewWithTag:2];
    rssi.text = [cur.discoveryTimeRSSI stringValue];
    
    UILabel *connected = (UILabel *)[cell viewWithTag:3];
    if (cur.state == CBPeripheralStateConnected) {
        [connected setHidden:NO];
    } else {
        [connected setHidden:YES];
    }
    
    UILabel *name = (UILabel *)[cell viewWithTag:4];
    name.text = cur.name;
    
    UIImageView *signal = (UIImageView *)[cell viewWithTag:5];
    if (cur.averageRSSI) {
        double movingAverage = cur.averageRSSI.doubleValue;
        if (movingAverage < -80.0) {
            signal.image = [UIImage imageNamed:@"wifi_d1"];
        } else if (movingAverage < -70.0) {
            signal.image = [UIImage imageNamed:@"wifi_d2"];
        } else if (movingAverage < -60.0) {
            signal.image = [UIImage imageNamed:@"wifi_d3"];
        } else if (movingAverage < -50.0) {
            signal.image = [UIImage imageNamed:@"wifi_d4"];
        } else if (movingAverage < -40.0) {
            signal.image = [UIImage imageNamed:@"wifi_d5"];
        } else {
            signal.image = [UIImage imageNamed:@"wifi_d6"];
        }
    } else {
        signal.image = [UIImage imageNamed:@"wifi_not_connected"];
    }
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
    
    self.selected = self.devices[indexPath.row];
    if (self.metaBootSwitch.on) {
        [self.scanningSwitch setOn:NO animated:YES];
        [self.metaBootSwitch setOn:NO animated:YES];
        [self metaBootSwitchPressed:self.metaBootSwitch];
        // Pause the screen while update is going on
        self.hud = [MBProgressHUD showHUDAddedTo:self.view animated:YES];
        self.hud.mode = MBProgressHUDModeDeterminateHorizontalBar;
        self.hud.label.text = @"Updating...";
        [[[self.selected prepareForFirmwareUpdateAsync] success:^(MBLFirmwareUpdateInfo * _Nonnull result) {
            DFUFirmware *selectedFirmware;
            if ([result.firmwareUrl.pathExtension caseInsensitiveCompare:@"zip"] == NSOrderedSame) {
                selectedFirmware = [[DFUFirmware alloc] initWithUrlToZipFile:result.firmwareUrl];
            } else {
                selectedFirmware = [[DFUFirmware alloc] initWithUrlToBinOrHexFile:result.firmwareUrl urlToDatFile:nil type:DFUFirmwareTypeApplication];
            }
            
            DFUServiceInitiator *initiator = [[DFUServiceInitiator alloc] initWithCentralManager:result.centralManager target:result.target];
            [initiator withFirmwareFile:selectedFirmware];
            initiator.forceDfu = YES; // We also have the DIS which confuses the DFU library
            initiator.logger = self; // - to get log info
            initiator.delegate = self; // - to be informed about current state and errors
            initiator.progressDelegate = self; // - to show progress bar
            initiator.peripheralSelector = self;
            
            [initiator start];
        }] failure:^(NSError * _Nonnull error) {
            NSLog(@"Firmware update error: %@", error.localizedDescription);
            [[[UIAlertView alloc] initWithTitle:@"Update Error"
                                        message:[@"Please re-connect and try again, if you can't connect, try MetaBoot Mode to recover.\nError: " stringByAppendingString:error.localizedDescription]
                                       delegate:nil
                              cancelButtonTitle:@"Okay"
                              otherButtonTitles:nil] show];
            [self.hud hideAnimated:YES];
        }];
    } else {
        [self performSegueWithIdentifier:@"DeviceDetails" sender:nil];
    }
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    return @"Devices";
}

#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    DeviceDetailViewController *destination = segue.destinationViewController;
    destination.device = self.selected;
}

#pragma mark - DFU Service delegate methods

- (void)didStateChangedTo:(enum State)state
{
    if (state == StateCompleted) {
        self.hud.mode = MBProgressHUDModeText;
        self.hud.label.text = @"Success!";
        [self.hud hideAnimated:YES afterDelay:2.0];
        [[MBLMetaWearManager sharedManager] clearDiscoveredDevices];
    }
}

- (void)didErrorOccur:(enum DFUError)error withMessage:(NSString *)message
{
    NSLog(@"Firmware update error %ld: %@", (long) error, message);
    [[[UIAlertView alloc] initWithTitle:@"Update Error"
                                message:[@"Please re-connect and try again, if you can't connect, try MetaBoot Mode to recover.\nError: " stringByAppendingString:message]
                               delegate:nil
                      cancelButtonTitle:@"Okay"
                      otherButtonTitles:nil] show];
    [self.hud hideAnimated:YES];
}

- (void)onUploadProgress:(NSInteger)part
             totalParts:(NSInteger)totalParts
               progress:(NSInteger)percentage
currentSpeedBytesPerSecond:(double)speed
 avgSpeedBytesPerSecond:(double)avgSpeed
{
    self.hud.progress = (double)percentage / 100.0;
}

- (void)logWith:(enum LogLevel)level message:(NSString *)message
{
    if (level >= LogLevelApplication) {
        NSLog(@"%d: %@", (int)level, message);
    }
}

#pragma mark - DFUPeripheralSelector

- (BOOL)select:(CBPeripheral *)peripheral advertisementData:(NSDictionary<NSString *, id> *)advertisementData RSSI:(NSNumber *)RSSI
{
    return [peripheral.identifier isEqual:self.selected.identifier];
}

- (NSArray<CBUUID *> *)filterBy
{
    return nil;
}

@end

File: MetaWearApiTest/DevicesTableViewController.swift
//
//  DevicesTableViewController.swift
//  MetaWearApiTest
//
//  Created by Stephen Schiffli on 11/2/16.
//  Copyright © 2016 MbientLab. All rights reserved.
//

import UIKit
import MetaWear
import MBProgressHUD
import iOSDFULibrary

class DevicesTableViewController: UITableViewController, DFUServiceDelegate, DFUProgressDelegate, LoggerDelegate, DFUPeripheralSelectorDelegate {
    var devices: [MBLMetaWear]?
    var activity: UIActivityIndicatorView!
    var hud: MBProgressHUD?
    var selected: MBLMetaWear?
    var initiator: DFUServiceInitiator?
    var dfuController: DFUServiceController?

    @IBOutlet weak var scanningSwitch: UISwitch!
    @IBOutlet weak var metaBootSwitch: UISwitch!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        activity = UIActivityIndicatorView(activityIndicatorStyle: .gray)
        activity.center = CGPoint(x: 95, y: 138)
        tableView.addSubview(activity)
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        setScanning(scanningSwitch.isOn)
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        setScanning(false)
    }
    
    func setScanning(_ on: Bool) {
        if on {
            activity.startAnimating()
            if metaBootSwitch.isOn {
                MBLMetaWearManager.shared().startScan(forMetaBootsAllowDuplicates: true, handler: { array in
                    self.devices = array
                    self.tableView.reloadData()
                })
            } else {
                MBLMetaWearManager.shared().startScan(forMetaWearsAllowDuplicates: true, handler: { array in
                    self.devices = array
                    self.tableView.reloadData()
                })
            }
        } else {
            activity.stopAnimating()
            MBLMetaWearManager.shared().stopScan()
        }
    }

    @IBAction func scanningSwitchPressed(_ sender: UISwitch) {
         setScanning(sender.isOn)
    }
    
    @IBAction func metaBootSwitchPressed(_ sender: Any) {
        MBLMetaWearManager.shared().stopScan()
        // Wait a split second for any final callbacks to fire before starting up scanning again
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
            self.devices = nil
            self.tableView.reloadData()
            self.setScanning(self.scanningSwitch.isOn)
        }
    }
    
    // MARK: - Table view data source
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return devices?.count ?? 0
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        let cur = devices![indexPath.row]
        
        let uuid = cell.viewWithTag(1) as! UILabel
        uuid.text = cur.identifier.uuidString
        
        let rssi = cell.viewWithTag(2) as! UILabel
        rssi.text = cur.discoveryTimeRSSI?.stringValue
        
        let connected = cell.viewWithTag(3) as! UILabel
        if cur.state == .connected {
            connected.isHidden = false
        } else {
            connected.isHidden = true
        }
        
        let name = cell.viewWithTag(4) as! UILabel
        name.text = cur.name
        
        let signal = cell.viewWithTag(5) as! UIImageView
        if let averageRSSI = cur.averageRSSI {
            let movingAverage = averageRSSI.doubleValue
            if movingAverage < -80.0 {
                signal.image = #imageLiteral(resourceName: "wifi_d1")
            } else if movingAverage < -70.0 {
                signal.image = #imageLiteral(resourceName: "wifi_d2")
            } else if movingAverage < -60.0 {
                signal.image = #imageLiteral(resourceName: "wifi_d3")
            } else if movingAverage < -50.0 {
                signal.image = #imageLiteral(resourceName: "wifi_d4")
            } else if movingAverage < -40.0 {
                signal.image = #imageLiteral(resourceName: "wifi_d5")
            } else {
                signal.image = #imageLiteral(resourceName: "wifi_d6")
            }
        } else {
            signal.image = #imageLiteral(resourceName: "wifi_not_connected")
        }
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return "Devices"
    }
    
    // MARK: - Table view delegate
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: false)
        selected = devices![indexPath.row]
        if (metaBootSwitch.isOn) {
            scanningSwitch.setOn(false, animated: true)
            metaBootSwitch.setOn(false, animated: true)
            metaBootSwitchPressed(metaBootSwitch)
            
            // Pause the screen while update is going on
            hud = MBProgressHUD.showAdded(to: view, animated: true)
            hud?.mode = .determinateHorizontalBar
            hud?.label.text = "Updating..."
            
            MBLMetaWearManager.shared().logLevel = .info
            selected?.prepareForFirmwareUpdateAsync().success { result in
                var selectedFirmware: DFUFirmware?
                if result.firmwareUrl.pathExtension.caseInsensitiveCompare("zip") == .orderedSame {
                    selectedFirmware = DFUFirmware(urlToZipFile: result.firmwareUrl)
                } else {
                    selectedFirmware = DFUFirmware(urlToBinOrHexFile: result.firmwareUrl, urlToDatFile: nil, type: .application)
                }
                self.initiator = DFUServiceInitiator(centralManager: result.centralManager, target: result.target)
                let _ = self.initiator?.with(firmware: selectedFirmware!)
                self.initiator?.forceDfu = true // We also have the DIS which confuses the DFU library
                self.initiator?.logger = self // - to get log info
                self.initiator?.delegate = self // - to be informed about current state and errors
                self.initiator?.peripheralSelector = self
                self.initiator?.progressDelegate = self // - to show progress bar
                
                self.dfuController = self.initiator?.start()
            }.failure { error in
                print("Firmware update error: \(error.localizedDescription)")
                let alertController = UIAlertController(title: "Update Error", message: "Please re-connect and try again, if you can't connect, try MetaBoot Mode to recover.\nError: \(error.localizedDescription)", preferredStyle: .alert)
                alertController.addAction(UIAlertAction(title: "Okay", style: .default, handler: nil))
                self.present(alertController, animated: true, completion: nil)
                self.hud?.hide(animated: true)
            }
            
        } else {
            performSegue(withIdentifier: "DeviceDetails", sender: nil)
        }
    }
    
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        let destination = segue.destination as! DeviceDetailViewController
        destination.device = selected
    }
    
    
    // MARK: - DFU Service delegate methods
    
    func dfuStateDidChange(to state: DFUState) {
        if state == .completed {
            hud?.mode = .text
            hud?.label.text = "Success!"
            hud?.hide(animated: true, afterDelay: 2.0)
            MBLMetaWearManager.shared().clearDiscoveredDevices()
        }
    }
    
    func dfuError(_ error: DFUError, didOccurWithMessage message: String) {
        print("Firmware update error \(error): \(message)")
        
        let alertController = UIAlertController(title: "Update Error", message: "Please re-connect and try again, if you can't connect, try MetaBoot Mode to recover.\nError: \(message)", preferredStyle: .alert)
        alertController.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
        present(alertController, animated: true, completion: nil)
        
        hud?.hide(animated: true)
    }
    
    func dfuProgressDidChange(for part: Int, outOf totalParts: Int, to progress: Int,
                              currentSpeedBytesPerSecond: Double, avgSpeedBytesPerSecond: Double) {
        hud?.progress = Float(progress) / 100.0
    }
    
    func logWith(_ level: LogLevel, message: String) {
        if level.rawValue >= LogLevel.application.rawValue {
            print("\(level): \(message)")
        }
    }
    func select(_ peripheral: CBPeripheral, advertisementData: [String : AnyObject], RSSI: NSNumber, hint name: String?) -> Bool {
        return peripheral.identifier == selected?.identifier
    }
    
    func filterBy(hint dfuServiceUUID: CBUUID) -> [CBUUID]? {
        return nil
    }
}
